Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -898,108 +897,121 @@ public final class ManifestLoader: ManifestLoaderProtocol {
-            try withTemporaryDirectory { tmpDir, cleanupTmpDir in
-                // Set path to compiled manifest executable.
-    #if os(Windows)
-                let executableSuffix = ".exe"
-    #else
-                let executableSuffix = ""
-    #endif
-                let compiledManifestFile = tmpDir.appending(component: "\(packageIdentity)-manifest\(executableSuffix)")
-                cmd += ["-o", compiledManifestFile.pathString]
-
-                // Compile the manifest.
-                Process.popen(arguments: cmd, environment: toolchain.swiftCompilerEnvironment, queue: delegateQueue) { result in
-                    var cleanupIfError = DelayableAction(target: tmpDir, action: cleanupTmpDir)
-                    defer { cleanupIfError.perform() }
-
-                    let compilerResult : ProcessResult
-                    do {
-                        compilerResult = try result.get()
-                        evaluationResult.compilerOutput = try (compilerResult.utf8Output() + compilerResult.utf8stderrOutput()).spm_chuzzle()
-                    } catch {
-                        return completion(.failure(error))
-                    }
-
-                    // Return now if there was an error.
-                    if compilerResult.exitStatus != .terminated(code: 0) {
-                        return completion(.success(evaluationResult))
-                    }
-
-                    // Pass an open file descriptor of a file to which the JSON representation of the manifest will be written.
-                    let jsonOutputFile = tmpDir.appending(component: "\(packageIdentity)-output.json")
-                    guard let jsonOutputFileDesc = fopen(jsonOutputFile.pathString, "w") else {
-                        return completion(.failure(StringError("couldn't create the manifest's JSON output file")))
-                    }
+        // wrap the completion with cleanup operations
+        let completion: (Result<EvaluationResult, Error>) -> Void = { result in
+            // free concurrency control semaphore
+            self.concurrencySemaphore.signal()
+            completion(result)
+        }
 
-                    cmd = [compiledManifestFile.pathString]
-        #if os(Windows)
-                    // NOTE: `_get_osfhandle` returns a non-owning, unsafe,
-                    // unretained HANDLE.  DO NOT invoke `CloseHandle` on `hFile`.
-                    let hFile: Int = _get_osfhandle(_fileno(jsonOutputFileDesc))
-                    cmd += ["-handle", "\(String(hFile, radix: 16))"]
-        #else
-                    cmd += ["-fileno", "\(fileno(jsonOutputFileDesc))"]
-        #endif
+        // we must not block the calling thread (for concurrency control) so nesting this in a queue
+        self.evaluationQueue.async {
+            do {
+                // park the evaluation thread based on the max concurrency allowed
+                self.concurrencySemaphore.wait()
+                // run the evaluation
```

Comment:
this diff is terrible but its basically nesting under the queue as noted in these 4 lines